-1 
Tested at: 416752306d7390cb48803e581c5db2d9ef7d5f1f
 I found an error when building: 
 
 <pre>>> Compiling edm plugin /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMOffline/Trigger/src/TopDiLeptonHLTOfflineDQM.cc 
>> Compiling edm plugin /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMOffline/Trigger/src/TopHLTOfflineDQMHelper.cc 
>> Compiling edm plugin /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMOffline/Trigger/src/TopSingleLeptonHLTOfflineDQM.cc 
>> Compiling edm plugin /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMOffline/Trigger/plugins/DQMGenericTnPClient.cc 
In file included from /tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMOffline/Trigger/plugins/DQMGenericTnPClient.cc:9:0:
/tmp/cmsbuild/workspace/ib-any-integration/CMSSW_8_0_X_2015-11-02-1100/src/DQMServices/Core/interface/DQMStore.h:9:36: fatal error: classlib/utils/Regexp.h: No such file or directory
 # include "classlib/utils/Regexp.h"
                                    ^
compilation terminated.
>> Building edm plugin tmp/slc6_amd64_gcc493/src/DQM/SiPixelMonitorCluster/src/DQMSiPixelMonitorCluster/libDQMSiPixelMonitorCluster.so
Leaving library rule at DQM/SiPixelMonitorCluster
</pre> 
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-12234/9426/summary.html

--- a/FWCore/Framework/interface/EDConsumerBase.h
+++ b/FWCore/Framework/interface/EDConsumerBase.h
@@ -75,8 +75,10 @@ namespace edm {
     typedef ProductLabels Labels;
     void labelsForToken(EDGetToken iToken, Labels& oLabels) const;
     
-    void modulesDependentUpon(const std::string& iProcessName,
-                              std::vector<const char*>& oModuleLabels) const;
+    void modulesDependentUpon(std::string const& iProcessName,
+                              std::string const& iModuleLabel,
+                              bool iPrint,
+                              std::vector<char const*>& oModuleLabels) const;
 
     void modulesWhoseProductsAreConsumed(std::vector<ModuleDescription const*>& modules,
                                          ProductRegistry const& preg,

--- a/FWCore/Framework/interface/Schedule.h
+++ b/FWCore/Framework/interface/Schedule.h
@@ -282,6 +282,7 @@ namespace edm {
     std::unique_ptr<SystemTimeKeeper> summaryTimeKeeper_;
 
     bool                           wantSummary_;
+    bool                           printDependencies_;
 
     volatile bool           endpathsAreActive_;
   };

--- a/FWCore/Framework/interface/stream/EDAnalyzerAdaptorBase.h
+++ b/FWCore/Framework/interface/stream/EDAnalyzerAdaptorBase.h
@@ -92,8 +92,10 @@ namespace edm {
       
       const EDConsumerBase* consumer() const;
       
-      void modulesDependentUpon(const std::string& iProcessName,
-                                std::vector<const char*>& oModuleLabels) const;
+      void modulesDependentUpon(std::string const& iProcessName,
+                                std::string const& iModuleLabel,
+                                bool iPrint,
+                                std::vector<char const*>& oModuleLabels) const;
 
       void modulesWhoseProductsAreConsumed(std::vector<ModuleDescription const*>& modules,
                                            ProductRegistry const& preg,

--- a/FWCore/Framework/interface/stream/ProducingModuleAdaptorBase.h
+++ b/FWCore/Framework/interface/stream/ProducingModuleAdaptorBase.h
@@ -82,8 +82,10 @@ namespace edm {
       void updateLookup(BranchType iBranchType,
                         ProductHolderIndexHelper const&);
 
-      void modulesDependentUpon(const std::string& iProcessName,
-                                std::vector<const char*>& oModuleLabels) const;
+      void modulesDependentUpon(std::string const& iProcessName,
+                                std::string const& iModuleLabel,
+                                bool iPrint,
+                                std::vector<char const*>& oModuleLabels) const;
 
       void modulesWhoseProductsAreConsumed(std::vector<ModuleDescription const*>& modules,
                                            ProductRegistry const& preg,

--- a/FWCore/Framework/src/EDConsumerBase.cc
+++ b/FWCore/Framework/src/EDConsumerBase.cc
@@ -20,6 +20,7 @@
 // user include files
 #include "FWCore/Framework/interface/EDConsumerBase.h"
 #include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/Utilities/interface/BranchType.h"
 #include "FWCore/Utilities/interface/Likely.h"
 #include "FWCore/Utilities/interface/Exception.h"
@@ -379,17 +380,25 @@ namespace {
 }
 
 void
-EDConsumerBase::modulesDependentUpon(const std::string& iProcessName,
-                                     std::vector<const char*>& oModuleLabels) const {
-  std::set<const char*, CharStarComp> uniqueModules;
-  for(unsigned int index=0, iEnd=m_tokenInfo.size();index <iEnd; ++index) {
+EDConsumerBase::modulesDependentUpon(std::string const& iProcessName,
+                                     std::string const& iModuleLabel,
+                                     bool iPrint,
+                                     std::vector<char const*>& oModuleLabels) const {
+  std::set<char const*, CharStarComp> uniqueModules;
+  for(unsigned int index = 0, iEnd = m_tokenInfo.size(); index < iEnd; ++index) {
     auto const& info = m_tokenInfo.get<kLookupInfo>(index);
-    if( not info.m_index.skipCurrentProcess() ) {
-      auto labels = m_tokenInfo.get<kLabels>(index);
-      unsigned int start = labels.m_startOfModuleLabel;
-      const char* processName = &(m_tokenLabels[start+labels.m_deltaToProcessName]);
-      if( (not processName) or processName[0]==0 or
-         iProcessName == processName) {
+    if(not info.m_index.skipCurrentProcess()) {
+      auto const& labels = m_tokenInfo.get<kLabels>(index);
+      unsigned int const start = labels.m_startOfModuleLabel;
+      char const* processName = &(m_tokenLabels[start+labels.m_deltaToProcessName]);
+      if(iPrint) {
+        LogAbsolute("ModuleDependency") << "<ModuleDependency> '" << iModuleLabel << 
+        "' may consume product of type '" << info.m_type.className() << 
+        "' with input tag '" << &(m_tokenLabels[start]) <<
+        ':' << &(m_tokenLabels[start+labels.m_deltaToProductInstance]) <<
+        ':' << processName << "'";;
+      }
+      if((not processName) or processName[0]==0 or iProcessName == processName) {
         uniqueModules.insert(&(m_tokenLabels[start]));
       }
     }

--- a/FWCore/Framework/src/Schedule.cc
+++ b/FWCore/Framework/src/Schedule.cc
@@ -347,6 +347,12 @@ namespace edm {
       proc_pset.addParameter<vstring>(std::string("@end_paths"), scheduledEndPaths);
       
     }
+
+    bool printDependencies(ParameterSet const& pset) {
+      ParameterSet defopts;
+      ParameterSet const& opts = pset.getUntrackedParameterSet("options", defopts);
+      return opts.getUntrackedParameter("printDependencies", false);
+    }
   }
   // -----------------------------
 
@@ -371,6 +377,7 @@ namespace edm {
     all_output_communicators_(),
     preallocConfig_(prealloc),
     wantSummary_(tns.wantSummary()),
+    printDependencies_(printDependencies(proc_pset)),
     endpathsAreActive_(true)
   {
     assert(0<prealloc.numberOfStreams());
@@ -400,7 +407,7 @@ namespace edm {
       std::copy(modulesToUse.begin(),itBeginUnscheduled,std::back_inserter(temp));
       temp.swap(modulesToUse);
     }
-    globalSchedule_.reset( new GlobalSchedule{ resultsInserter_,
+    globalSchedule_.reset(new GlobalSchedule{ resultsInserter_,
       moduleRegistry_,
       modulesToUse,
       proc_pset, preg, prealloc,
@@ -409,7 +416,7 @@ namespace edm {
     //TriggerResults is not in the top level ParameterSet so the call to
     // reduceParameterSet would fail to find it. Just remove it up front.
     std::set<std::string> usedModuleLabels;
-    for( auto const worker: allWorkers()) {
+    for(auto const worker: allWorkers()) {
       if(worker->description().moduleLabel() != kTriggerResults) {
         usedModuleLabels.insert(worker->description().moduleLabel());
       }
@@ -1279,8 +1286,8 @@ namespace edm {
   {
     //Need to lookup names to ids quickly
     std::map<std::string,unsigned int> moduleNamesToIndex;
-    for(auto worker: allWorkers()) {
-      moduleNamesToIndex.insert( std::make_pair(worker->description().moduleLabel(),
+    for(auto const& worker: allWorkers()) {
+      moduleNamesToIndex.insert(std::make_pair(worker->description().moduleLabel(),
                                          worker->description().id()));
     }
     
@@ -1307,8 +1314,8 @@ namespace edm {
           auto found = alreadySeenNames.insert(name);
           if(found.second) {
             //first time for this path
-            unsigned int moduleIndex = moduleNamesToIndex[name];
-            if(not lastModuleName.empty() ) {
+            unsigned int const moduleIndex = moduleNamesToIndex[name];
+            if(not lastModuleName.empty()) {
               edgeToPathMap[std::make_pair(moduleIndex,lastModuleIndex)].push_back(pathIndex);
             }
             lastModuleName = name;
@@ -1324,14 +1331,14 @@ namespace edm {
       for(auto const& worker: allWorkers()) {
         dependentModules.clear();
         //NOTE: what about aliases?
-        worker->modulesDependentUpon(dependentModules);
+        worker->modulesDependentUpon(dependentModules, printDependencies_);
         auto found = moduleNamesToIndex.find(worker->description().moduleLabel());
         if (found == moduleNamesToIndex.end()) {
           //The module was from a previous process
           continue;
         }
-        unsigned int moduleIndex = found->second;
-        for(auto name: dependentModules) {
+        unsigned int const moduleIndex = found->second;
+        for(auto const& name: dependentModules) {
           edgeToPathMap[std::make_pair(moduleIndex, moduleNamesToIndex[name])].push_back(std::numeric_limits<unsigned int>::max());
         }
       }

--- a/FWCore/Framework/src/Worker.h
+++ b/FWCore/Framework/src/Worker.h
@@ -108,7 +108,7 @@ namespace edm {
     virtual void updateLookup(BranchType iBranchType,
                       ProductHolderIndexHelper const&) = 0;
 
-    virtual void modulesDependentUpon(std::vector<const char*>& oModuleLabels) const = 0;
+    virtual void modulesDependentUpon(std::vector<const char*>& oModuleLabels, bool iPrint) const = 0;
 
     virtual void modulesWhoseProductsAreConsumed(std::vector<ModuleDescription const*>& modules,
                                                  ProductRegistry const& preg,

--- a/FWCore/Framework/src/WorkerT.h
+++ b/FWCore/Framework/src/WorkerT.h
@@ -109,8 +109,8 @@ namespace edm {
     virtual void implRegisterThinnedAssociations(ProductRegistry const&, ThinnedAssociationsHelper&) override;
     virtual std::string workerType() const override;
 
-    virtual void modulesDependentUpon(std::vector<const char*>& oModuleLabels) const override {
-      module_->modulesDependentUpon(module_->moduleDescription().processName(),oModuleLabels);
+    virtual void modulesDependentUpon(std::vector<const char*>& oModuleLabels, bool iPrint) const override {
+      module_->modulesDependentUpon(module_->moduleDescription().processName(),module_->moduleDescription().moduleLabel(), iPrint, oModuleLabels);
     }
 
     virtual void modulesWhoseProductsAreConsumed(std::vector<ModuleDescription const*>& modules,

--- a/FWCore/Framework/src/stream/EDAnalyzerAdaptorBase.cc
+++ b/FWCore/Framework/src/stream/EDAnalyzerAdaptorBase.cc
@@ -114,10 +114,12 @@ EDAnalyzerAdaptorBase::consumer() const {
 }
 
 void
-EDAnalyzerAdaptorBase::modulesDependentUpon(const std::string& iProcessName,
-                                            std::vector<const char*>& oModuleLabels) const {
+EDAnalyzerAdaptorBase::modulesDependentUpon(std::string const& iProcessName,
+                                            std::string const& iModuleLabel,
+                                            bool iPrint,
+                                            std::vector<char const*>& oModuleLabels) const {
   assert(not m_streamModules.empty());
-  return m_streamModules[0]->modulesDependentUpon(iProcessName, oModuleLabels);
+  return m_streamModules[0]->modulesDependentUpon(iProcessName, iModuleLabel, iPrint, oModuleLabels);
 }
 
 void

--- a/FWCore/Framework/src/stream/ProducingModuleAdaptorBase.cc
+++ b/FWCore/Framework/src/stream/ProducingModuleAdaptorBase.cc
@@ -106,10 +106,12 @@ namespace edm {
 
     template< typename T>
     void
-    ProducingModuleAdaptorBase<T>::modulesDependentUpon(const std::string& iProcessName,
-                                                        std::vector<const char*>& oModuleLabels) const {
+    ProducingModuleAdaptorBase<T>::modulesDependentUpon(std::string const& iProcessName,
+                                                        std::string const& iModuleLabel,
+                                                        bool iPrint,
+                                                        std::vector<char const*>& oModuleLabels) const {
       assert(not m_streamModules.empty());
-      return m_streamModules[0]->modulesDependentUpon(iProcessName, oModuleLabels);
+      return m_streamModules[0]->modulesDependentUpon(iProcessName, iModuleLabel, iPrint, oModuleLabels);
     }
 
     template< typename T>

--- a/FWCore/Framework/test/BuildFile.xml
+++ b/FWCore/Framework/test/BuildFile.xml
@@ -306,3 +306,7 @@
   <flags TEST_RUNNER_ARGS=" /bin/bash FWCore/Framework/test test_earlyTerminationSignal.sh"/>
   <use name="FWCore/Utilities"/>
 </bin>
+<bin   name="TestFWCoreFrameworkPrintDependencies" file="TestDriver.cpp">
+  <flags   TEST_RUNNER_ARGS=" /bin/bash FWCore/Framework/test run_PrintDependencies.sh"/>
+  <use name="FWCore/Utilities"/>
+</bin>

--- a/FWCore/Framework/test/run_PrintDependencies.sh
+++ b/FWCore/Framework/test/run_PrintDependencies.sh
@@ -1,0 +1,12 @@
+#!/bin/bash
+
+# Pass in name and status
+function die { echo $1: status $2 ;  exit $2; }
+
+F1=${LOCAL_TEST_DIR}/testPrintDependencies.py
+
+cmsRun $F1 2>&1 | grep "may consume" >& ${LOCAL_TEST_DIR}/dependencies.txt || die "Failure using $F1" $?
+diff -q ${LOCAL_TEST_DIR}/dependencies.txt ${LOCAL_TEST_DIR}/unit_test_outputs/dependencies.txt || die "dependencies differ" $?
+
+rm -f ${LOCAL_TEST_DIR}/dependencies.txt
+

--- a/FWCore/Framework/test/testPrintDependencies.py
+++ b/FWCore/Framework/test/testPrintDependencies.py
@@ -1,0 +1,25 @@
+import FWCore.ParameterSet.Config as cms
+
+process = cms.Process("TESTOUTPUT")
+process.load("FWCore.Framework.test.cmsExceptionsFatal_cff")
+
+process.options= cms.untracked.PSet(
+    printDependencies = cms.untracked.bool(True)
+)
+process.maxEvents = cms.untracked.PSet(
+    input = cms.untracked.int32(1)
+)
+process.Thing = cms.EDProducer("ThingProducer")
+
+process.OtherThing = cms.EDProducer("OtherThingProducer")
+
+process.output = cms.OutputModule("PoolOutputModule",
+    fileName = cms.untracked.string('file:PoolOutputTest.root')
+)
+
+process.source = cms.Source("EmptySource")
+
+process.p = cms.Path(process.Thing*process.OtherThing)
+process.ep = cms.EndPath(process.output)
+
+

--- a/FWCore/Framework/test/unit_test_outputs/dependencies.txt
+++ b/FWCore/Framework/test/unit_test_outputs/dependencies.txt
@@ -1,0 +1,8 @@
+<ModuleDependency> 'OtherThing' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing::'
+<ModuleDependency> 'output' may consume product of type 'edm::TriggerResults' with input tag 'TriggerResults::TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::OtherThing>' with input tag 'OtherThing:testUserTag:TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing::TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing:beginLumi:TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing:beginRun:TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing:endLumi:TESTOUTPUT'
+<ModuleDependency> 'output' may consume product of type 'std::vector<edmtest::Thing>' with input tag 'Thing:endRun:TESTOUTPUT'

